File: kotlin/bob/_iterations/2/src/main/kotlin/Bob.kt
object Bob {
    fun hey(input: String): String = if (input.silence) {
        "Fine. Be that way!"
    } else when (Pair(input.question, input.yelling)) {
        Pair(true, true) -> "Calm down, I know what I'm doing!"
        Pair(true, false) -> "Sure."
        Pair(false, true) -> "Whoa, chill out!"
        else -> "Whatever."
    }

    private val String.silence: Boolean
        get() = all { it.isWhitespace() }

    private val String.question: Boolean
        get() = trim().endsWith('?')

    private val String.yelling: Boolean
        get() = any { it.isUpperCase() } && none { it.isLowerCase() }
}

File: ruby/bob/_iterations/2/bob.rb
module Bob
  def self.hey(remark)
    return "Fine. Be that way!" if silent?(remark)
    
    case [question?(remark), yelling?(remark)]
      when [true, true]
        "Calm down, I know what I'm doing!"
      when [false, true]
        "Whoa, chill out!"
      when [true, false]
        "Sure."
      else
        "Whatever."
    end
  end

  def self.silent?(remark)
    !!(/\A\s*\Z/ =~ remark)
  end

  def self.question?(remark)
    !!(/\?\s*\Z/ =~ remark)
  end

  def self.yelling?(remark)
    !!(/\A[^a-z]*[A-Z][^a-z]*\Z/ =~ remark)
  end
end
